client/src/procedures/primitives.rs

```diff
@@ -789,3 +798,115 @@ impl UseSecret for AeadDecrypt {
         &self.key
     }
 }
+
+/// Executes the concat KDF as defined in Section 5.8.1 of NIST.800-56A.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ConcatKdf {
+    /// The hash function to use in the kdf.
+    pub hash: Sha2Hash,
```

Comment:
Sha-2 Family based hashes are susceptible to [length extension attacks ](https://github.com/marcelo140/length-extension). I would advise to take another digest. blake2b would be a good change. 

Or is SHA256 / X required for this algorithm? I couldn't find an explicit mentioning [here](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf)

Reply:
The SHA256 digest is specified by the JWA RFC 7518 for ECDH: https://www.rfc-editor.org/rfc/rfc7518.html#section-4.6.2

> Key derivation is performed using the Concat KDF, as defined in
>   Section 5.8.1 of [[NIST.800-56A](https://www.rfc-editor.org/rfc/rfc7518.html#ref-NIST.800-56A)], where the Digest Method is SHA-256.

Reply:
Correct me if I'm wrong, but my understanding of the length extension attack is that:

- It requires the original, valid hash to perform. Here that would be the output of the Concat KDF itself, which is intended to be a shared secret and thus, if compromised, would mean the attacker would be able to decrypt any related messages anyway, without performing a length extension attack. 
- It still results in a different hash, and thus would not compromise the derived key here.

The above does make the assumption that the output of the Concat KDF is not used as a MAC, but that seems to be the case for all intended purposes.

I agree that SHA-2 is not desirable, but given that it is specified as a requirement in RFC 7518, we still need to use it to be interoperable with existing implementations.

Reply:
agreed. Looks like there is no other way around.